src/ddmd/root/filename.d

```diff
@@ -721,3 +722,112 @@ nothrow:
+
+        // different conventions for CreateDirectory and mkdir
+        return createRet == 0 ? 1 : 0;
+    }
+
+    // Converts a path to one suitable to be passed to Win32 API
+    // functions that can deal with paths longer than 248
+    // characters then calls the supplied function on it.
+    // For more information:
+    // https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx
+    package auto extendedPathThen(alias F)(const(char*) path)
+    {
+        import core.sys.windows.winbase: GetFullPathNameW;
+
+        wchar[1024] wpathBuf;
+        // Since the unicode Win32 APIs use UTF16, we need to convert
+        // from UTF8 to UTF16 first.
+        const wpath = path.toWStringz(wpathBuf);
+
+        // Now wpath is the UTF16 version of path, but to be able to use
+        // extended paths, we need to prefix with `\\?\` and the absolute
+        // path.
+        static immutable prefix = `\\?\`w;
+
+        // We don't know how large the absolute file path will be.
+        // We need to tell GetFullPathNameW the size of the buffer we're
+        // passing in so we defensively allocate 4 times the size of
+        // the string we have currently.
+        const absLength = prefix.length + wpath.length * 4;
+        static wchar[1024] absBuf;
+        auto absPath = absLength > absBuf.length ? new wchar[absLength] : absBuf[];
+
+        absPath[0 .. prefix.length] = prefix[];
+
+        const absPathRet = GetFullPathNameW(&wpath[0],
+                                            absPath.length,
+                                            &absPath[prefix.length],
+                                            null /*filePartBuffer*/);
+
+        auto extendedPath = absPathRet > absPath.length ? null : absPath[0 .. absPathRet];
+        return F(extendedPath);
+    }
+
+
+    // Converts an UTF8 null-terminated string to an array of wchar that's null
+    // terminated so it can be passed to Win32 APIs.
+    // buf is passed as a scratch space to store the result. If more memory
+    // is needed then toWstringz allocates on the GC heap instead.
+    private wchar[] toWStringz(const(char*) str, wchar[] buf = []) pure nothrow
+    {
+        import core.stdc.string: strlen;
+
+        const length8 = strlen(str);
+
+        // Conservative estimate that should be enough for any UTF8 string
+        // according to https://tools.ietf.org/html/rfc3629.
+        // The +1 is for the null terminator.
+        const length16 = length8 * 4 + 1;
+
+        auto wstr = length16 > buf.length ? new wchar[length16] : buf;
+
+        // Using i, wchar c in the foreach doesn't work since then i would
+        // be tied to the length of the UTF8 sequence.
+        int wstrLen;
+        try
+            foreach(wchar c; str[0 .. length8]) wstr[wstrLen++] = c;
+        catch(Exception)
+            return null;
+
+        wstr[wstrLen] = 0; // null-terminate it
+
+        return wstr[0 .. wstrLen];
+    }
```

Comment:
can't use https://msdn.microsoft.com/en-us/library/windows/desktop/dd319072(v=vs.85).aspx ?

Reply:
What would I pass as the code page?

Reply:
`CP_UTF8`? The magical transcoding loop (iterating `char[]` by `wchar`s) you're using a few lines above is assuming that anyway, right?

Reply:
Still wondering about this.

Reply:
I'm not convinced it'd be any simpler - there would be two calls to that function (the first to find out the needed space to allocate) along with associated calls to `GetLastError`.